server/websocket.go

```diff
@@ -667,190 +685,278 @@ func (p *Plugin) handleJoin(userID, connID, authSessionID string, joinData Calls
-			}, clusterMessageTypeConnect, handlerID); err != nil {
-				return fmt.Errorf("failed to send connect message: %w", err)
+		if batcher == nil {
+			batchMaxSize := min(int(channelStats.MemberCount), maxJoinLeaveOpsBatchSize)
+			p.LogDebug("creating new addSessionsBatcher for call", "channelID", channelID, "batchMaxSize", batchMaxSize)
+			batcher, err = batching.NewBatcher(batching.Config{
+				Interval: joinLeaveBatchingInterval,
+				Size:     batchMaxSize,
+				PreRunCb: func(ctx batching.Context) error {
+					p.LogDebug("performing addSessionToCall batch", "channelID", channelID, "batchSize", ctx[batching.ContextBatchSizeKey])
+
+					state, err := p.lockCallReturnState(channelID)
+					if err != nil {
+						return fmt.Errorf("failed to lock call: %w", err)
+					}
+					ctx["callState"] = state
+					return nil
+				},
+				PostRunCb: func(_ batching.Context) error {
+					p.unlockCall(channelID)
+					return nil
+				},
+			})
+			if err != nil {
+				return fmt.Errorf("failed to create batcher: %w", err)
 			}
+			p.addSessionsBatchers[channelID] = batcher
+			batcher.Start()
 		}
-	}
-
-	// send successful join response
-	p.publishWebSocketEvent(wsEventJoin, map[string]interface{}{
-		"connID": connID,
-	}, &WebSocketBroadcast{UserID: userID, ReliableClusterSend: true})
-
-	if len(state.sessionsForUser(userID)) == 1 {
-		// Only send event on first session join.
-		// This is to keep backwards compatibility with clients not supporting
-		// multi-sessions.
-		p.publishWebSocketEvent(wsEventUserConnected, map[string]interface{}{
-			"userID": userID,
-		}, &WebSocketBroadcast{ChannelID: channelID, ReliableClusterSend: true})
-	}
 
-	p.publishWebSocketEvent(wsEventUserJoined, map[string]interface{}{
-		"user_id":    userID,
-		"session_id": connID,
-	}, &WebSocketBroadcast{ChannelID: channelID, ReliableClusterSend: true})
-
-	if userID == p.getBotID() && state.Recording != nil {
-		p.publishWebSocketEvent(wsEventCallJobState, map[string]interface{}{
-			"callID":   channelID,
-			"jobState": getClientStateFromCallJob(state.Recording).toMap(),
-		}, &WebSocketBroadcast{
-			ChannelID:           channelID,
-			ReliableClusterSend: true,
-			UserIDs:             getUserIDsFromSessions(state.sessions),
+		err = batcher.Push(func(ctx batching.Context) {
+			ctx["callState"] = addSessionToCall(ctx["callState"].(*callState))
 		})
+		if err != nil {
+			return fmt.Errorf("failed to push to batcher: %w", err)
+		}
 
-		// MM-57224: deprecated, remove when not needed by mobile pre 2.14.0
-		p.publishWebSocketEvent(wsEventCallRecordingState, map[string]interface{}{
-			"callID":   channelID,
-			"recState": getClientStateFromCallJob(state.Recording).toMap(),
-		}, &WebSocketBroadcast{
-			ChannelID:           channelID,
-			ReliableClusterSend: true,
-			UserIDs:             getUserIDsFromSessions(state.sessions),
-		})
+		return nil
 	}
 
-	clientStateData, err := json.Marshal(state.getClientState(p.getBotID(), userID))
+	// Non-batching case
```

Comment:
From the community data and load tests do you have any idea how much a problem quick-joining is? I can see quick-ending  being a legit problem, but I don't think I've seen a call where people joined all at once. 
That's two separate questions I guess: 1) is quick-joining a serious technical problem, and 2) is quick-joining a real-life problem.

Reply:
I think it is as soon as you scale up due to the nature of meetings. They are planned in advance, have a fixed date/time and people will naturally join around that time. If you have enough people (e.g. hundreds) they will likely ramp up very quickly in the first couple of minutes.